source/common/thread_local/thread_local_impl.cc

```diff
@@ -60,6 +99,32 @@ void InstanceImpl::registerThread(Event::Dispatcher& dispatcher, bool main_threa
     dispatcher.post([&dispatcher] { thread_local_data_.dispatcher_ = &dispatcher; });
   }
 }
+void InstanceImpl::recycle(std::unique_ptr<SlotHolder>&& holder) {
+  if (holder->isRecycleable()) {
+    holder.reset();
+    return;
+  }
+  deferred_deletes_.emplace_back(std::move(holder));
+  scheduleCleanup();
+}
+
+void InstanceImpl::scheduleCleanup() {
+  ASSERT(main_thread_dispatcher_ != nullptr);
+  if (shutdown_) {
+    return;
+  }
+  if (deferred_deletes_.empty()) {
+    return;
+  }
+  main_thread_dispatcher_->post([this]() {
+    deferred_deletes_.remove_if(
+        [](std::unique_ptr<SlotHolder>& holder) -> bool { return holder->isRecycleable(); });
+    if (!deferred_deletes_.empty()) {
```

Comment:
Prefer a loop vs. recursion.

Reply:
Let me update the misleading comment.
This method will reschedule another cleanup task in the main thread dispatcher Q. Not exactly recursive.
We can't use a loop here as it basically means a lock/wait mechanism (on the ongoing update callbacks been finished).

Reply:
Thanks this makes more sense now. Can you add a bunch more comments inside `recycle` and this function to describe what is going on and the various interleavings?

Reply:
Ping on adding more comments.

Reply:
oops, missed this. 
I added some more comments around the function header. and also recycle.